### DISCUSSION TITLE
Best practices against brute-force attack

### DISCUSSION THREAD

**Discussion by markovendelin** [minio/minio#16018] [Q&A]
I am considering using LDAP for our MinIO users authentication. However, I wonder what is the recommended practice to avoid brute-force attacks against MinIO service. There was an issue regarding use of fail2ban with MinIO (https://github.com/minio/minio/issues/12503), but it would be tricky as we use IPVLAN docker driver and fail2ban wouldn't have any control over that. In addition, looks like fail2ban is not trivial (if possible) to setup in MinIO context (see the issue).

So, I wonder whether 
- there is some internal mechanism that would count failed attempts on MinIO side and will prohibit next ones on the basis of some threshold (similar to fail2ban)?
- maybe it is possible to have additional authentication on top of LDAP, as TOTP or similar?

As an LDAP backend, we use FreeIPA. If we could have OAuth2 for physical users accessing the service via web, creating their service accounts there and then would use service accounts to access via S3. Such solution should also protect from brute-force attacks, as far as I understand.

Or maybe some other solution?

**Comment by dfloyd888** [comment]
It would be nice to have some tarpit or timeout functionality where a host doing active hammering can be set aside for a period of time, just to slow down the rate of attacks.

I think part of the reason why brute force attack mitigation is not an issue is that for the average AWS key, there is a large random name for the user, and a large password, which makes trying to guess that infeasible, so much that attacks pretty much won't succeed.

What can be done as a workaround would be to have some anti-brute force protection on the FreeIPA side.  You can also have users have built in 2FA authentication (where they type their password plus their Google Authenticator code in the password field), which would help.
